crates/bevy_pbr/src/meshlet/cull_clusters.wgsl

```diff
@@ -148,14 +140,22 @@ fn cull_clusters(
     meshlet_raster_clusters[buffer_slot] = cluster_id;
 }
 
-// https://stackoverflow.com/questions/21648630/radius-of-projected-sphere-in-screen-space/21649403#21649403
-fn lod_error_is_imperceptible(cp: vec3<f32>, r: f32) -> bool {
-    let d2 = dot(cp, cp);
-    let r2 = r * r;
-    let sphere_diameter_uv = view.clip_from_view[0][0] * r / sqrt(d2 - r2);
-    let view_size = f32(max(view.viewport.z, view.viewport.w));
-    let sphere_diameter_pixels = sphere_diameter_uv * view_size;
-    return sphere_diameter_pixels < 1.0;
+// TODO: This doesn't account for perspective and other edge cases as well as Nanite does
+// https://github.com/zeux/meshoptimizer/blob/1e48e96c7e8059321de492865165e9ef071bffba/demo/nanite.cpp#L115
+fn lod_error_is_imperceptible(lod_sphere: MeshletBoundingSphere, simplification_error: f32, world_from_local: mat4x4<f32>, world_scale: f32) -> bool {
+    let sphere_world_space = (world_from_local * vec4(lod_sphere.center, 1.0)).xyz;
+    let radius_world_space = world_scale * lod_sphere.radius;
+    let error_world_space = world_scale * simplification_error;
+
+    let distance_to_closest_point_on_sphere = distance(sphere_world_space, view.world_position) - radius_world_space;
+    let distance_to_closest_point_on_sphere_clamped_to_znear = max(distance_to_closest_point_on_sphere, view.clip_from_view[3][2]);
+    var projected_error = error_world_space / distance_to_closest_point_on_sphere_clamped_to_znear;
+    if view.clip_from_view[3][3] != 1.0 {
```

Comment:
Is this an ortho check? If so I think that should be done separately; in an ortho projection you don't need to divide by distance at all (but you do still need to multiply by ortho scaling similar to perspective)

Reply:
Yep, this is an orthographic projection check. Thanks for letting me know, I'll take another stab at trying to understand the theory and figure out what to do for Ortho.

Reply:
Is this correct? Not sure I applied the ortho scaling factor correctly

```rust
var projected_error = error_world_space;
if view.clip_from_view[3][3] != 1.0 {
    // Perspective
    let distance_to_closest_point_on_sphere = distance(sphere_world_space, view.world_position) - radius_world_space;
    let distance_to_closest_point_on_sphere_clamped_to_znear = max(distance_to_closest_point_on_sphere, view.clip_from_view[3][2]);
    projected_error *= view.clip_from_view[1][1] / distance_to_closest_point_on_sphere_clamped_to_znear;
} else {
    // Orthographic
    projected_error *= view.clip_from_view[1][1] * 0.5;
}
projected_error *= view.viewport.w;
```

Where the ortho matrix is
```rust
pub fn orthographic_rh(
    left: f32,
    right: f32,
    bottom: f32,
    top: f32,
    near: f32,
    far: f32,
) -> Self {
    let rcp_width = 1.0 / (right - left);
    let rcp_height = 1.0 / (top - bottom);
    let r = 1.0 / (near - far);
    Self::from_cols(
        Vec4::new(rcp_width + rcp_width, 0.0, 0.0, 0.0),
        Vec4::new(0.0, rcp_height + rcp_height, 0.0, 0.0),
        Vec4::new(0.0, 0.0, r, 0.0),
        Vec4::new(
            -(left + right) * rcp_width,
            -(top + bottom) * rcp_height,
            r * near,
            1.0,
        ),
    )
}```

Reply:
I think you’d want 0.5 factor for perspective as well? Ortho math seems correct (just eyeballing)

Reply:
Oh yeah, I missed that when copying your code. Thanks for the extra pair of eyes :)

Reply:
Reduces down to this:
```rust
var projected_error = error_world_space;
if view.clip_from_view[3][3] != 1.0 {
    // Perspective
    let distance_to_closest_point_on_sphere = distance(sphere_world_space, view.world_position) - radius_world_space;
    let distance_to_closest_point_on_sphere_clamped_to_znear = max(distance_to_closest_point_on_sphere, view.clip_from_view[3][2]);
    projected_error *= view.clip_from_view[1][1] / distance_to_closest_point_on_sphere_clamped_to_znear;
}
projected_error *= view.clip_from_view[1][1] * 0.5;
projected_error *= view.viewport.w;
```